File: 22-generate-parentheses/generate-parentheses.ts
function generateParenthesis(n: number): string[] {
    const ans = [];

    const dfs = (oc: number, cc: number, cur: string): void => {
        if (oc === cc && cur.length === n * 2) {
            ans.push(cur);
            return;
        }
        if (oc < n) {
            
            dfs(oc + 1, cc, cur + "(");
        }
        if (cc < oc) {
            dfs(oc, cc + 1, cur + ")");
        }
    }
    dfs(0, 0, "");
    return ans;
};

File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    List<String> ans = new ArrayList<>();
    public List<String> generateParenthesis(int n) {
        generate("(", 1, n);
        return ans;
    }

    private void generate(String curStr, int openCount , int n) {
        int m = curStr.length();
        if(m == n * 2){
            ans.add(curStr);
            return;
        }
        int closedCount = m - openCount;
        if(openCount < n) generate(curStr + "(" , openCount + 1 , n);
        if(closedCount < n && closedCount < openCount) generate(curStr + ")" , openCount , n);
    }
}